File: 백준/Gold/13549. 숨바꼭질 3/숨바꼭질 3.py
import sys
import queue
input = sys.stdin.readline
INF = int(1e9)
MAXN = 100000
N, K = map(int, input().split())
visited = [False] * (MAXN + 1)
q = queue.Queue()
q.put((N,0))
ret = INF
retCount = 0
time = 0
while not q.empty():
    current, time = q.get()
    if current < 0 or current > MAXN:
        continue
    visited[current] = True
    if current == K:
        if ret > time:
            ret = time
            retCount = 1
        elif ret == time:
            retCount += 1
        continue
    if (current - 1) >= 0 and not visited[current - 1]:
        q.put(((current - 1), time + 1))
    if current < K and (current + 1) <= MAXN and not visited[current + 1]:
        q.put(((current + 1), time + 1))
    if current < K and (current * 2) <= MAXN and not visited[current * 2]:
        q.put(((current * 2), time))
print(ret)

File: 백준/Gold/13549. 숨바꼭질 3/숨바꼭질 3.java
import java.io.*;
import java.util.*;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static int ret;
    static int N, M;
    static int In[];
    static boolean visited[];
    static Queue<Jump> q = new LinkedList<>();
    public static void main(String[] args)    throws IOException {
        In = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        N = In[0];
        M = In[1];
        visited = new boolean[100001];
        ret = Integer.MAX_VALUE;
        q.offer(new Jump(In[0], 0));
        while (!q.isEmpty()) {
            Jump j = q.poll();
            visited[j.x] = true;
            if (j.x == M) {
                ret = Math.min(ret, j.cnt);
                continue;
            } else if (j.x < 0 || j.x > 100000) {
                continue;
            }
            if (j.x < M) {
                if (!visited[j.x+1])
                    q.offer(new Jump(j.x+1, j.cnt+1));
                if (j.x*2 <= 100000 && !visited[j.x*2])
                    q.offer(new Jump(j.x*2, j.cnt));
                if (j.x-1 >= 0 && !visited[j.x-1])
                    q.offer(new Jump(j.x-1, j.cnt+1));
            } else {
                if (j.x-1 >= 0 && !visited[j.x-1])
                    q.offer(new Jump(j.x-1, j.cnt+1));
            }
        }
        bw.write(ret + "\n");
        bw.flush();
        bw.close();
    }
    public static class Jump {
        int x;
        int cnt;
        public Jump(int x, int cnt) {
            this.x = x;
            this.cnt = cnt;
        }
    }
}